examples/simple_fwd_tb/forward.c

```diff
@@ -0,0 +1,331 @@
+/*********************************************************************
+ *                     openNetVM
+ *              https://sdnfv.github.io
+ *
+ *   BSD LICENSE
+ *
+ *   Copyright(c)
+ *            2020 National Institute of Technology Karnataka, Surathkal
+ *   All rights reserved.
+ *
+ *   Redistribution and use in source and binary forms, with or without
+ *   modification, are permitted provided that the following conditions
+ *   are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *     * The name of the author may not be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *   OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * forward.c - an example using onvm. Simulates a queue with a Token Bucket
+ * and forwards packets to a DST NF.
+ ********************************************************************/
+
+#include <errno.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <unistd.h>
+
+#include <rte_common.h>
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+#include <rte_cycles.h>
+
+#include "onvm_nflib.h"
+#include "onvm_pkt_helper.h"
+
+#define NF_TAG "simple_fwd_tb"
+
+#define PKT_READ_SIZE ((uint16_t) 100000000) // TODO: Find proper value for PKT_READ_SIZE
```

Comment:
We usually set this to 32. See the NF scaling_example and speed_tester.

Reply:
Thanks for the feedback. I wanted to know if there is a specific reason we set it to 32?
In this NF, we need to determine if a packet would fit in the given depth or if it was surplus. To do that, I'm dequeuing all packets available in the rx_ring (hence the large value). This way, I can easily determine if I need to process the packet or drop it.
If we dequeue 32 packets at once everytime, it would be hard to determine whether the packets needs to be dropped or processed.

Reply:
Also, I just realized that uint16_t can't hold the given large value. Will fix that soon.

Reply:
@rohit-mp I haven't looked at the code yet, but I'm curious how your approach handles the case where the number of packets in the ring at each lookup tends to be small, but added together they might exceed your specified depth.  Do you handle this case?

The rings are relatively small. If I remember right, the size is defined as 512 here: 
https://github.com/sdnfv/openNetVM/blob/46bbc962a0ef2ddfd774a7fda798f9ea92b7b116/onvm/onvm_mgr/onvm_init.h#L90-L91

Reply:
Thanks for pointing out the max size of the ring @twood02 . I think that this implies a upper limit on the depth that can be supplied to the NF.

The difference between my approach and a traditional token bucket is that I dequeue packets in a burst instead of one-by-one, and the depth constraint is not enforced at the queue (rx_ring) but is taken care elsewhere. 
After each burst dequeue operation, I process enough packets that fulfill the depth constraint and drop the rest. For this, I try to dequeue as many packets as possible from the ring, so that I know exactly which packets were excess, and the ring becomes empty after each dequeue operation.

If the number of packets in the ring at each lookup is small, it would mean that the rate of token production is high enough to allow the packets to be processed fast enough (assuming that slow rate of packet production is not the reason). In this case, I think it's fine that together they would exceed the depth since they're being processed fast enough such that the depth constraint is not violated at any point of time.

I would be happy to know your opinion on this.

Reply:
In simple turns, i think of a token bucket working something like: 1) release packets as long as you have tokens, 2) every second increase your number of tokens.  It sounds like you are treating a call to dequeue as if it is your period for resetting the number of tokens.  That will give an approximation of the token bucket algorithm, but it won't be precise since we don't know how many packets are arriving between each dequeue or how much time passes between each dequeue.  For example, with the same parameters for your token bucket you'll get different results depending on how fast the system is.

This may be OK for your purposes, I just want to be sure you are thinking about how these factors will influence your output rate.  (And this will be a great example for helping people learn how to use the advanced rings interface in any case!)

Reply:
Just to make calculations easier, instead of increasing tokens every second, I increase them only when they are insufficient while keeping track of last increased time.
I treat my call to dequeue as a reset on the queue size (I try to dequeue all the packets available in the ring at that moment). This helps in identifying excess packets easily.

But yes, this is still an approximation of the token bucket algorithm since the packets which would've been dropped as compared between this implementation and dequeuing one-by-one would be different. This was a sacrifice I had to make to implement this without maintaining a separate queue of my own. Another approach I thought of was maintaining another queue and having two threads, one dequeuing from rx_ring and enqueuing to my queue (while taking care of the depth) and another dequeuing from my queue and enqueuing to tx_ring, but this would lead to maintaining locks on the queue which could lead to a performance drop.

Reply:
ah ok, that makes more sense. Yes, if you do your timing checks at the time of adjusting the tokens then I think this is reasonable.

Reply:
But this approach relies on emptying the ring at each dequeue. Any suggestions on what to set the value of `PKT_READ_SIZE` to be? I think 512 * max_pkt_size would be sufficient, or I could remove that constant altogether and check the size of ring just before dequeue and use that value.

Reply:
Thanks for the clarification. Yea, I think that in this case checking the size then dequeueing is the way to go.

Reply:
> The rings are relatively small. If I remember right, the size is defined as 512 here:
> 
> https://github.com/sdnfv/openNetVM/blob/46bbc962a0ef2ddfd774a7fda798f9ea92b7b116/onvm/onvm_mgr/onvm_init.h#L90-L91

Thanks for the feedback. On reducing the size of the array of packet pointers (to hold the dequeued packets) to 512 and dequeuing all the packets from the ring, the NF started to crash. So I checked the capacity of the ring and saw it to be 16383. So I did a bit of searching and found this code which matches with what I saw: 
https://github.com/sdnfv/openNetVM/blob/440ac6b28ac91d6e9e68e67f9465001dcf77c4af/onvm/onvm_nflib/onvm_common.h#L70
I see that this value is being passed as count in creating the rings for the NF.
Can you help me understand this? Thanks, again!

Reply:
That value, `RTE_MP_*_DESC_DEFAULT` appears to be related to the port configuration. `NF_QUEUE_RINGSIZE` is the size of the RX/TX ring. You may either check the size and dequeue all packets up to that maximum, or just set the dequeue burst parameter to the ringsize. I think the latter approach is more efficient.

Reply:
Setting the dequeue burst parameter to `NF_QUEUE_RINGSIZE` works like a charm. Thanks!